전역 에러코드/예외 정의 및 예외 처리 작성

street62 opened the pull request:
### Summary
- 앱 내에서 사용될 전역 에러코드와 각 에러코드에 해당하는 예외 클래스들을 정의하고, 실행 도중 예외가 발생했을 때 이를 처리하는 핸들러들을 작성했습니다.

### Description
**ErrorCode 및 커스텀 예외 정의**
- ErrorCode의 경우에는 크게 `400(BAD REQUEST)`, `401(UNAUTHORIZED)`, `403(FORBIDDEN)`, `404(NOT FOUND)`로 나누어 작성했습니다.
- 각 예외마다 `ErrorCode`에 해당하는 메시지를 갖도록 했고, 401에 해당하는 `AuthorizationFailedException`의 경우에는 해당하는 에러 코드가 여러 개여서 세부적으로 ErrorCode를 인자로 받아 메시지를 정할 수 있도록 했습니다. 혹시 더 필요한/불필요한 ErrorCode가 있다면 의견 남겨주세요!

**컨트롤러 예외 핸들링**
- 컨트롤러 단으로 올라오는 예외는 `@ControllerAdvice`를 통해 처리하도록 했습니다.

**스프링 시큐리티 예외 핸들링**
- 스프링 시큐리티 필터 내에서 JWT 검증 시 발생하는 예외는 `JwtProvider`에서 로그 처리 이후 `AuthorizationFailedException`으로 변환하여 던지도록 하고, 해당 예외를 `AuthorizationExceptionFilter`에서 잡아 처리하도록 했습니다.
- 이 구조 이해가 조금 어려웠는데, 디버거를 통해서 확인해 본 바로는 `AuthorizationFailedException`은 정의한 `AuthorizationExceptionFilter`의 catch문을 통해서 잡히고, 이외의 `RuntimeException`들은 계속 타고 올라가다가 `ExceptionTranslationFilter`에서 `AuthenticationException`으로 변환되어 `AuthenticationEntryPoint`에서 잡히는 구조인 듯합니다. 

**스프링 시큐리티 Config 의존성 구조 변경**
- `SecurityConfig`에서 각 필터 체인들을 작성할 때 기존에는 `@Bean` 메서드를 통해서 생성 후 주입받고 있었는데요. 필터가 추가될 때 필터 생성에 필요한 인자들이 계속 `SecurityConfig`의 필드로 추가되는 것이 조금 걸려서, 각 Filter 클래스 파일에서 `@Component`로 처리한 후 `SecurityConfig`에서는 해당 필터들을 생성자로 주입받는 형태로 변경해 보았습니다. 혹시 이 부분에 대해서도 의견 있으시면 말씀해 주세요.

시큐리티 복잡한데... 신기하고 재밌기도 하네요 흐흐
진행하면서 계속 같이 공부해 보면 좋을 것 같아요!

street62 committed:
✨ 전역 ErrorCode enum 정의

HTTP 상태코드 기준 400, 401, 403, 404로 나누어 정의.

street62 committed:
✨ 커스텀 예외 정의

- 인증 관련 예외인 AuthorizationFailedException은 기본 생성자로 생성 시 UNAUTHORIZED 에러코드 메시지를 넣고, 이외의 경우에는 ErrorCode를 추가로 정의할 수 있도록 함
- 이외 예외는 기본 생성자에서 매칭되는 ErrorCode를 사용하도록 함

street62 committed:
✨ 글로벌 예외 핸들러 정의

각 예외에 맞는 응답을 ResponseEntity 형식으로 보내도록 함

street62 committed:
🎨 예외 패키지 정리

common/exceptions 내부에 /handler와 /errorcode 추가

street62 committed:
🎨 RestAuthenticationEntryPoint 응답 수정

status code와 메시지를 보내도록 수정

street62 committed:
🎨 WebMvcConfig AUTHORIZATION_HEADER 오타 수정

'Autorization' -> 'Authorization'으로 수정

street62 committed:
✨ AuthorizationExceptionFilter 작성

인증 관련 예외 발생 시 ErrorCode에 지정된 메시지 반환

street62 committed:
🎨 JwtTokenProvider 토큰 검증 예외 AuthorizationFailedException으로 변경해 반환

- 기존 프린트문을 log로 교체
- 토큰 검증 과정에서 발생하는 예외들을 AuthorizationFailedException으로 변경하여 던지도록 함

street62 committed:
🎨 JwtAuthenticationFilter 컴포넌트로 등록

SecurityConfig에서 커스텀 필터 사용 시 Bean의 필드들을 주입해주어야 한다는 점을 고려해 각 Filter를 클래스 레벨에서 `@Component`로 선언하도록 변경.

street62 committed:
🎨 SecurityConfig 필터 의존성 변경, 필터체인 임시 변경

- JwtAuthenticationFilter, AuthorizationExceptionFilter 필드로 주입받도록 함
- 테스트를 위해 RestAuthenticationEntryPoint 임시 삭제. 추후 수정 예정.

street62 committed:
🎨 SecurityConfig authenticationEntryPoint 복구

- SecurityConfig에서 커스텀 필터 선언 이후 exceptionHandling().authenticationEntryPoint() 복구
- 커스텀 예외인 AuthorizationFailedException은 authorizationExceptionFilter에서 처리하고, 이외 RuntimeException은 ExceptionTranslationFilter에서 AuthenticationException으로 변환 후 처리하는 것으로 확인

RokwonK commented on src/main/java/com/nexters/keyme/auth/filter/JwtAuthenticationFilter.java:
```diff
@@ -21,6 +22,7 @@
 
 
 @Slf4j
+@Component
 @RequiredArgsConstructor
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
 
```
위 OncePerRequestFilter 내용과 동일합니다. 수정부탁드립니다.

RokwonK replied:
아이코 코멘트 순서가 바뀌었군요 OncePerRequestFilter에 관한 설명은 아래 코멘트에 있습니다!

RokwonK commented on src/main/java/com/nexters/keyme/auth/util/JwtTokenProvider.java:
```diff
@@ -88,15 +90,17 @@ private Claims checkTokenValidation(String token, JwtParser parser) {
+      log.info("Invalid JWT token");
+      throw new AuthorizationFailedException(ErrorCode.TOKEN_INVALID);
     } catch (ExpiredJwtException e) {
-      System.out.println("Expired JWT token");
+      log.info("Expired JWT token");
+      throw new AuthorizationFailedException(ErrorCode.TOKEN_EXPIRED);
     } catch (UnsupportedJwtException e) {
-      System.out.println("Unsupported JWT token");
+      log.info("Unsupported JWT token");
+      throw new AuthorizationFailedException(ErrorCode.TOKEN_INVALID);
     } catch (IllegalArgumentException e) {
-      System.out.println("JWT claims string is empty.");
+      log.info("JWT claims string is empty.");
+      throw new AuthorizationFailedException(ErrorCode.UNAUTHORIZED);
```
👍

RokwonK commented on src/main/java/com/nexters/keyme/auth/filter/AuthorizationExceptionFilter.java:
```diff
@@ -0,0 +1,34 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Component
+@RequiredArgsConstructor
+public class AuthorizationExceptionFilter extends OncePerRequestFilter {
+
+    private final ObjectMapper objectMapper;
```
OncePerRequestFilter는 Filter 인터페이스를 구현한 추상클래스입니다. 따라서 Component로 등록시에 Security Filter가 아닌 Filter로 등록되는 문제가 일어날 수 있습니다. 따라서 Component로 등록해주시면 안됩니다.

RokwonK replied:
SecurityConfig에서 주입받아서 사용하면 FIlter와 Security Filter 모두에 등록되겠군요

street62 replied:
오옷 이런 부분도 있군요! 모르면 나중에 큰 문제가 생길수도 있겠네요. 감사합니다.

https://bitgadak.tistory.com/m/10

`Filter를 상속하고 스프링 컨테이너의 빈으로 등록되는 클래스들은 ApplicationFilterChain에도 추가된다`는 흐름이면 `@Bean`도 결국 메서드 리턴값을 컴포넌트로 등록하니까 같은 문제가 발생하지 않을까 싶은데 혹시 차이점이 있을까요?

대신 위 블로그에서 마지막 부분에 한 것처럼 컴포넌트나 빈 애노테이션 없이 new 키워드로 필터 클래스의 생성자 호출해서 넣는 로직이 되어야 하지 않을까 하는 생각이 드는데, 제가 놓친 부분이 있으면 말씀해주세요!

RokwonK replied:
> 'Bean도 결국 메서드 리턴값을 컴포넌트로 등록하니까 같은 문제가 발생하지 않을까 싶은데 혹시 차이점이 있을까요?' 

당시에는 Configuration 내부에서 작성하니까 문제 없는건가? 라는 얼토당토 않은 생각을 했는데 다시생각하니 말씀하신대로 차이가 없는 것 같습니다. 조금 검색도 해봤는데 확실한 답은 찾을 수 없었습니다. 
정확하지 않은 지식으로 혼란을 드려 죄송합니다😂

street62 replied:
앗 아님다 아님다 그럼 그냥 빈 등록 없이 생성자 넣어서 필터 넣는 걸로 변경할게요!

RokwonK replied:
넵 감사합니다!👍

RokwonK commented on src/main/java/com/nexters/keyme/config/SecurityConfig.java:
```diff
@@ -21,9 +20,8 @@
-
-  private final JwtTokenProvider jwtTokenProvider;
-  private final AuthenticationTokenProvider authenticationTokenProvider;
+  private final JwtAuthenticationFilter jwtAuthenticationFilter;
+  private final AuthorizationExceptionFilter authorizationExceptionFilter;
```
요것도 주입말고 Bean 메서드로 생성해주세요!

RokwonK commented on src/main/java/com/nexters/keyme/config/WebMvcConfig.java:
```diff
@@ -14,7 +12,7 @@
 public class WebMvcConfig implements WebMvcConfigurer {
 
-    public static final String AUTHORIZATION_HEADER = "Autorization";
+    public static final String AUTHORIZATION_HEADER = "Authorization";
```
와우 오타수정 감사합니다!

RokwonK reviewed (commented):
수고하셨습니다~ 아래 코멘트 확인해주시고 수정부탁드리겠습니다!

RokwonK committed:
🎨 EXPIRATION_MS long 으로 타입 변경

RokwonK commented:
추가적으로 토큰만료일 관련된 변수 타입을 long으로 변경하고 이 브랜치에 푸시했습니다!

street62 committed:
🎨 SecurityConfig 필터 추가 시 생성자 호출하도록 변경

- 기존에 사용하던 빈 등록 후 주입 방식은 ApplicationFilterChain에도 필터가 추가되는 문제 존재
- 빈 등록 관련 애노테이션 제거 후 생성자 호출해 넣는 방식으로 변경

street62 commented:
시큐리티 설정 관련 내용 수정 완료했습니다!
